chain/chain/src/flat_storage_resharder.rs

```diff
@@ -396,8 +405,74 @@ impl FlatStorageResharder {
         // Terminate the resharding event.
         *self.resharding_event.lock().unwrap() = None;
     }
+
+    /// Returns an iterator over a shard's flat storage at the given block hash. This
+    /// iterator contains both flat storage values and deltas.
+    fn flat_storage_iterator<'a>(
+        &self,
+        flat_store: &'a FlatStoreAdapter,
+        shard_uid: &ShardUId,
+        block_hash: &CryptoHash,
+    ) -> Result<Box<FlatStorageAndDeltaIter<'a>>, Error> {
+        let mut iter: Box<FlatStorageAndDeltaIter<'a>> = Box::new(
+            flat_store
+                .iter(*shard_uid)
+                // Get the flat storage iter and wrap the value in Optional::Some to
+                // match the delta iterator so that they can be chained.
+                .map_ok(|(key, value)| (key, Some(value)))
+                // Wrap the iterator's item into an Entry.
+                .map(|entry| FlatStorageAndDeltaIterItem::Entry(entry)),
+        );
+
+        // Get all the blocks from flat head to the wanted block hash.
+        let flat_storage = self
```

Comment:
I'm a bit confused about having flat_store as an argument and getting flat_storage from runtime here. Why do you need both and how do they differ? Is there any way to make it simpler? Ideally there would only be one flat `thing` or they would be clearly different and both would be obtained same way.

Reply:
Yes, it's a bit weird. I need `flat_store` passed to the function due to lifetimes on the iterator - this prevents me from instantiating a `flat_store` inside the `flat_storage_iterator`.

And then `flat_storage` and `flat_store`, although they have similar name, are totally independent. I couldn't get one from the other